torchaudio/functional.py

```diff
@@ -75,6 +76,8 @@ def spectrogram(
         ``n_fft // 2 + 1`` and ``n_fft`` is the number of
         Fourier bins, and time is the number of window hops (n_frame).
     """
+    if power is not None and return_complex:
+        raise ValueError('When `power` is provided, `return_complex` must be False.')
```

Comment:
why is that?

Reply:
In the original implementation, when the `power` argument is provided, the returned signal changes from pseudo-complex Tensor to plain real Tensor. (as the spectrogram is passed to `complex_norm` function which is equivalent of `abs` of complex type tensor).

We could still return complex type with all the imaginary parts equal zero, but I thought there is not much point of that. I am open to changing it though.

Reply:
`torch.abs` returns a floating point tensor though so I don't see any reason why we should change it

Reply:
However for performance reasons we should add a special case for power 1 (not call pow), and pow 2 (check if torch.view_as_real(x).pow(2).sum(-1) is performance wise better that `x.abs().pow(2)`).

Reply:
Okay, thanks.

> However for performance reasons we should add a special case for power 1 (not call pow), and pow 2 (check if torch.view_as_real(x).pow(2).sum(-1) is performance wise better that `x.abs().pow(2)`).

I will look into this.

Reply:
I got the following result, `torch.view_as_real(x).pow(2).sum(-1)` is slightly faster on CUDA.

| Method                                            | CPU                                   | GPU                              |
|---------------------------------|------------------------------------------|------------------------------------------|
| `x.abs().pow(2)`                                          | 100 loops, best of 5: 3.17 msec per loop | 100 loops, best of 5: 48.4 usec per loop |
| `torch.view_as_real(x).pow(2).sum(-1)` | 100 loops, best of 5: 28.8 msec per loop | 100 loops, best of 5: 41.7 usec per loop |

<details> <summary>script</summary>

```
OMP_NUM_THREADS=1 numactl --membind 0 --cpubind 0 python -m timeit -n 100 -r 5 -s """
import torch;
import torch.fft;
x = torch.zeros(1, 32*44100, dtype=torch.cfloat)
""" """
x.abs().pow(2);
"""

OMP_NUM_THREADS=1 numactl --membind 0 --cpubind 0 python -m timeit -n 100 -r 5 -s """
import torch;
import torch.fft;
x = torch.zeros(1, 32*44100, dtype=torch.cfloat)
""" """
torch.view_as_real(x).pow(2).sum(-1)
"""

OMP_NUM_THREADS=1 numactl --membind 0 --cpubind 0 python -m timeit -n 100 -r 5 -s """
import torch;
import torch.fft;
x = torch.zeros(1, 32*44100, dtype=torch.cfloat, device='cuda')
""" """
x.abs().pow(2);
"""

OMP_NUM_THREADS=1 numactl --membind 0 --cpubind 0 python -m timeit -n 100 -r 5 -s """
import torch;
import torch.fft;
x = torch.zeros(1, 32*44100, dtype=torch.cfloat, device='cuda')
""" """
torch.view_as_real(x).pow(2).sum(-1)
"""
```

</details>

Reply:
@anjali411 Updated the code based on the result.